calyx/src/ir/component.rs

```diff
@@ -103,3 +98,84 @@ impl Component {
         self.namegen.gen_name(prefix)
     }
 }
+
+/// A wrapper struct exposing an ordered collection of named entities within an
+/// RRC with deterministic iteration and constant-time look-up on names
+/// directly. The struct assumes that the name of an entity cannot change. Doing
+/// so will introduce incorrect results for look-ups.
+#[derive(Debug)]
+pub struct IdList<T: GetName>(LinkedHashMap<Id, RRC<T>>);
+
+impl<T: GetName> IdList<T> {
+    /// Removes all elements from the collection
+    pub fn clear(&mut self) {
+        self.0.clear();
+    }
+
+    /// Keep only the elements in the collection which satisfy the given
+    /// predicate
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&RRC<T>) -> bool,
+    {
+        for entry in self.0.entries() {
+            if !f(entry.get()) {
+                entry.remove();
+            }
+        }
+    }
+
+    /// Add a new element to the colleciton
+    pub fn add(&mut self, item: RRC<T>) {
+        let name = item.clone_name();
+        self.0.insert(name, item);
+    }
+
+    /// Add multiple elements to the collection from an owned vector
+    pub fn add_multiple(&mut self, items: Vec<RRC<T>>) {
+        for item in items {
+            self.add(item)
+        }
+    }
+
+    /// Add multiple elements to the collection from a slice. This will create
+    /// new clones of the given RRCs.
+    pub fn add_multiple_by_ref(&mut self, items: &[RRC<T>]) {
```

Comment:
Is this used anywhere?

Reply:
not currently, but someone had made a comment about not needing RRC ownership so I added this